lib/active_merchant/billing/gateways/xpay.rb

```diff
@@ -15,44 +15,59 @@ class XpayGateway < Gateway
 
       ENDPOINTS_MAPPING = {
         purchase: 'orders/2steps/payment',
+        authorize: 'orders/2steps/payment',
         preauth: 'orders/2steps/init',
-        capture: 'operations/{%s}/captures',
+        capture: 'operations/%s/captures',
         verify: 'orders/card_verification',
-        void: 'operations/{%s}/cancels',
-        refund: 'operations/{%s}/refunds'
+        void: 'operations/%s/cancels',
+        refund: 'operations/%s/refunds'
       }
 
+      module PaymentActions
+        PREAUTH = :preauth
+        AUTHORIZE = :authorize
+        PURCHASE = :purchase
+        CAPTURE = :capture
+        VERIFY = :verify
+        VOID = :void
+        REFUND = :refund
+      end
+
       def initialize(options = {})
         requires!(options, :api_key)
         @api_key = options[:api_key]
         super
       end
 
+      def preauth(amount, payment_method, options = {})
+        post = {}
+        add_transaction_params_commit(:preauth, amount, post, payment_method, options)
+      end
+
       def purchase(amount, payment_method, options = {})
         post = {}
-        add_auth_purchase_params(post, amount, payment_method, options)
-        action = options[:operation_id] ? :purchase : :preauth
-        commit(action, post, options)
+        add_transaction_params_commit(:purchase, amount, post, payment_method, options)
       end
 
       def authorize(amount, payment_method, options = {})
         post = {}
-        add_auth_purchase_params(post, amount, payment_method, options)
-        commit(:preauth, post, options)
+        add_transaction_params_commit(:authorize, amount, post, payment_method, options)
       end
 
       def capture(amount, authorization, options = {})
         post = {}
+        add_refund_capture_params(amount, post, options)
         commit(:capture, post, options)
       end
 
       def void(authorization, options = {})
-        post = {}
+        post = { description: options[:description] }
         commit(:void, post, options)
       end
 
-      def refund(amount, authorization, options = {})
+      def refund(amount, options = {})
```

Comment:
`authorization` is a part of the `refund` method signature, so it's a required argument

Reply:
Understood! Corrected!